finance_dl/schwab.py

```diff
@@ -0,0 +1,295 @@
+        output_directory: str,
+        min_start_date: datetime.date,
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.credentials = credentials
+        self.output_directory = output_directory
+        self.min_start_date = min_start_date
+        self.already_got_accounts: Set[Account] = set()
+        self.current_page = PageType.NONE
+
+    def run(self) -> None:
+        self.load_history_page()
+
+    def load_history_page(self) -> None:
+        logger.info("Loading history page.")
+        self.driver.get(self.HISTORY_URL)
+
+        self.login_if_needed()
+        self.current_page = PageType.HISTORY
+
+        seen: Set[Account] = set()
+
+        while True:
+            account = self.select_next_unseen_account(seen)
+            if account is None:
+                break
+            seen.add(account)
+            self.download(account)
+
+    def download(self, account: Account) -> None:
+        assert self.current_page == PageType.HISTORY
+
+        logger.info(f"Checking account {account}")
+
+        account_dir, positions_dir = self.get_account_dirs(account)
+        from_date = self.get_last_fetched_date(account_dir)
+        if from_date is None:
+            from_date = self.min_start_date
+        else:
+            from_date += self.ONE_DAY
+        # Only download up to yesterday, so we can avoid overlap and not risk missing
+        # any transactions.
+        to_date = datetime.date.today() - self.ONE_DAY
+        if to_date <= from_date:
+            logger.info("No dates to download.")
+            return
+
+        logger.info("Downloading transactions.")
+
+        num_transaction_types = self.get_num_transaction_types()
+        transaction_filter = "|".join(map(str, range(num_transaction_types)))
+
+        from_str = from_date.strftime("%m/%d/%Y")
+        to_str = to_date.strftime("%m/%d/%Y")
+        self.driver.get(
+            "https://client.schwab.com/api/History/Brokerage/ExportTransaction"
+            f"?sortSeq=1&sortVal=0&tranFilter={transaction_filter}"
+            f"&timeFrame=0&filterSymbol=&fromDate={from_str}&toDate={to_str}"
+            "&exportError=&invalidFromDate=&invalidToDate=&symbolExportValue="
+            "&includeOptions=N&displayTotal=true"
+        )
+        dest_name = f"{from_date.strftime('%Y-%m-%d')}_{to_date.strftime('%Y-%m-%d')}.csv"
+        dest_path = os.path.join(account_dir, dest_name)
+        self.get_file(f"{account.label}_Transactions_", dest_path)
+
+        logger.info("Downloading positions.")
+
+        self.driver.get(
+            "https://client.schwab.com/api/PositionV2/PositionsDataV2/Export"
+            "?CalculateDayChangeIntraday=true"
+            "&firstColumn=symbolandDescriptionStacked&format=csv"
+        )
+        dest_name = datetime.date.today().strftime("%Y-%m-%d") + ".csv"
+        dest_path = os.path.join(positions_dir, dest_name)
+        self.get_file(f"{account.label}-Positions-", dest_path)
```

Comment:
Instead of get_file, you could use `self.wait_and_return(self.get_downloaded_file)`.  However, since the download appears to be via URL rather than link, you could instead specify use_seleniumrequests=True to super.__init__ and then just use `self.driver.request('GET', url)`.  Refer to paypal.py for an example.

Reply:
I switched from my hand-rolled sleep loop to using `self.wait_and_return` -- agree that's definitely better, thanks!

I still kinda prefer a custom `get_file` over using `get_downloaded_file`, because it allows me to verify that the downloaded file matches the expected filename, rather than just implicitly assuming that any non-partial file found in the download dir must be the one we want. Also I don't really see a reason to read the file contents into the process when all I actually want to do is move it to a new location and name.

If you have a strong preference for `use_seleniumrequests` I could try that, but it feels more fragile to me because it seems like it requires setting the request headers explicitly? Or maybe those are just overrides, and it still defaults to whatever headers the browser would set? In any case the current code works and I don't see a strong motivation for switching to `use_seleniumrequests`.

Reply:
seleniumrequests attempts to replicate the headers (including cookies) that the browser would send normally --- that is the whole point of it.  In general anything that involves selenium tends to be less reliable and slower than making requests directly, but if you are happy with the current method that is certainly fine too.

Note that the download_dir is a temporary directory created just for this session, so if there is more than one downloaded file that would mean some previous action caused a file to be downloaded, which would suggest something has gone wrong.  get_downloaded_file does also return the file name, which you could check, but in general unless you need to extract some information from the file name (e.g. a date, which might otherwise be hard to get) I would suggest to just ignore the filename to make your code less likely to break if the website changes.